build(docs-infra): turn on Ivy

gkalpak opened the pull request:
The angular.io project uses Angular and CLI v9, which by default turns on Ivy mode. However, since ec4381d, we explicitly opt out of Ivy.

This commit removes the `enabledIvy: false` configuration, thus allowing the default behavior of having Ivy on.

NOTE:
This commit only changes the angular.io projects. The docs examples need to be updated separately (first to Angular and CLI v9 and then to Ivy).

##
Jira issue: [FW-413](https://angular-team.atlassian.net/browse/FW-413)
(I also created [FW-1607](https://angular-team.atlassian.net/browse/FW-1607) to track updating the docs examples to latest `@angular/*` and Ivy.)

---
I haven't profiled this, but the performance feels on par with ViewEngine when using the app.

There is currently a difference in payload sizes however. The lazy-loaded bundles are a little smaller with Ivy, but the main (eagerly loaded) bundle (and thus the overall app size) has increased (see below for details).

> _EDIT: Updating to CLI v9.0.0-next.6 improves things a lot for Ivy (bringing the different to ~2%, but there also seems to be a ~12KB regression in in the ES2015 main bundle for ViewEngine._ 🤔 

<details>
  <summary><b>Sizes with cli@9.0.0-next.4</b> (~10% increase with Ivy)</summary>

<p align="right"><sub><i>* Bold numbers indicate an undesirable change.</i></sub></p>

**ES2015**

| File | VE (B) | IVY (B) | Diff (B) | Diff (%) |
| --- | ---: | ---: | ---: | ---: |
| **_Eager bundles_** |
| `main` | 436411 | 496036 | **59625** | **14%** |
| `polyfills` | 52503 | 52503 | _0_ | _0%_ |
| `runtime` | 3097 | 2987 | _-110_ | _-4%_ |
| **_Total_** | 492011 | 551526 | **59515** | **12%** |
| &nbsp; | &nbsp; | &nbsp; | &nbsp; | &nbsp; |
| **_Lazy bundles_** |
| `announcement-bar-announcement-bar-module` | 2632 | 1848 | _-784_ | _-30%_ |
| `api-api-list-module` | 8349 | 6261 | _-2088_ | _-25%_ |
| `code-code-example-module` | 3120 | 2237 | _-883_ | _-28%_ |
| `code-code-tabs-module` | 35930 | 33545 | _-2385_ | _-7%_ |
| `contributor-contributor-list-module` | 7417 | 5327 | _-2090_ | _-28%_ |
| `default~code-code-example-module~code-code-tabs-module` | 52597 | 53171 | **574** | **1%** |
| `live-example-live-example-module` | 5239 | 4270 | _-969_ | _-18%_ |
| `resource-resource-list-module` | 4702 | 3460 | _-1242_ | _-26%_ |
| `search-file-not-found-search-module` | 1567 | 1213 | _-354_ | _-23%_ |
| `toc-toc-module` | 6373 | 5209 | _-1164_ | _-18%_ |
| **_Total_** | 127926 | 116541 | _-11385_ | _-9%_ |
| &nbsp; | &nbsp; | &nbsp; | &nbsp; | &nbsp; |
| **_TOTAL_** | 619937 | 668067 | **48130** | **8%** |

---
<p align="right"><sub><i>* Bold numbers indicate an undesirable change.</i></sub></p>

**ES5**

| File | VE (B) | IVY (B) | Diff (B) | Diff (%) |
| --- | ---: | ---: | ---: | ---: |
| **_Eager bundles_** |
| `main` | 481663 | 541663 | **60000** | **12%** |
| `polyfills` | 142239 | 142239 | _0_ | _0%_ |
| `runtime` | 3091 | 2981 | _-110_ | _-4%_ |
| **_Total_** | 626993 | 686883 | **59890** | **10%** |
| &nbsp; | &nbsp; | &nbsp; | &nbsp; | &nbsp; |
| **_Lazy bundles_** |
| `announcement-bar-announcement-bar-module` | 2767 | 1954 | _-813_ | _-29%_ |
| `api-api-list-module` | 9084 | 6942 | _-2142_ | _-24%_ |
| `code-code-example-module` | 3536 | 2614 | _-922_ | _-26%_ |
| `code-code-tabs-module` | 38140 | 35244 | _-2896_ | _-8%_ |
| `contributor-contributor-list-module` | 7616 | 5431 | _-2185_ | _-29%_ |
| `default~code-code-example-module~code-code-tabs-module` | 57173 | 57641 | **468** | **1%** |
| `live-example-live-example-module` | 5548 | 4525 | _-1023_ | _-18%_ |
| `resource-resource-list-module` | 4838 | 3548 | _-1290_ | _-27%_ |
| `search-file-not-found-search-module` | 1603 | 1218 | _-385_ | _-24%_ |
| `toc-toc-module` | 6778 | 5567 | _-1211_ | _-18%_ |
| **_Total_** | 137083 | 124684 | _-12399_ | _-9%_ |
| &nbsp; | &nbsp; | &nbsp; | &nbsp; | &nbsp; |
| **_TOTAL_** | 764076 | 811567 | **47491** | **6%** |

</details>

---
<details>
  <summary><b>Sizes with cli@9.0.0-next.6</b> (~2% increase with Ivy)</summary>

<p align="right"><sub><i>* Bold numbers indicate an undesirable change.</i></sub></p>

**ES2015**

| File | VE (B) | IVY (B) | Diff (B) | Diff (%) |
| --- | ---: | ---: | ---: | ---: |
| **_Eager bundles_** |
| `main` | 448708 | 458422 | **9714** | **2%** |
| `polyfills` | 52608 | 52608 | _0_ | _0%_ |
| `runtime` | 3106 | 2996 | _-110_ | _-4%_ |
| **_Total_** | 504422 | 514026 | **9604** | **2%** |
| &nbsp; | &nbsp; | &nbsp; | &nbsp; | &nbsp; |
| **_Lazy bundles_** |
| `announcement-bar-announcement-bar-module` | 2632 | 1848 | _-784_ | _-30%_ |
| `api-api-list-module` | 8357 | 6265 | _-2092_ | _-25%_ |
| `code-code-example-module` | 3120 | 2237 | _-883_ | _-28%_ |
| `code-code-tabs-module` | 35998 | 33582 | _-2416_ | _-7%_ |
| `contributor-contributor-list-module` | 7445 | 5341 | _-2104_ | _-28%_ |
| `default~code-code-example-module~code-code-tabs-module` | 52860 | 53411 | **551** | **1%** |
| `live-example-live-example-module` | 5247 | 4270 | _-977_ | _-19%_ |
| `resource-resource-list-module` | 4702 | 3460 | _-1242_ | _-26%_ |
| `search-file-not-found-search-module` | 1567 | 1213 | _-354_ | _-23%_ |
| `toc-toc-module` | 6395 | 5223 | _-1172_ | _-18%_ |
| **_Total_** | 128323 | 116850 | _-11473_ | _-9%_ |
| &nbsp; | &nbsp; | &nbsp; | &nbsp; | &nbsp; |
| **_TOTAL_** | 632745 | 630876 | _-1869_ | _0%_ |

---
<p align="right"><sub><i>* Bold numbers indicate an undesirable change.</i></sub></p>

**ES5**

| File | VE (B) | IVY (B) | Diff (B) | Diff (%) |
| --- | ---: | ---: | ---: | ---: |
| **_Eager bundles_** |
| `main` | 482744 | 501393 | **18649** | **4%** |
| `polyfills` | 142383 | 142383 | _0_ | _0%_ |
| `runtime` | 3100 | 2990 | _-110_ | _-4%_ |
| **_Total_** | 628227 | 646766 | **18539** | **3%** |
| &nbsp; | &nbsp; | &nbsp; | &nbsp; | &nbsp; |
| **_Lazy bundles_** |
| `announcement-bar-announcement-bar-module` | 2767 | 1954 | _-813_ | _-29%_ |
| `api-api-list-module` | 9092 | 6946 | _-2146_ | _-24%_ |
| `code-code-example-module` | 3536 | 2614 | _-922_ | _-26%_ |
| `code-code-tabs-module` | 38206 | 35279 | _-2927_ | _-8%_ |
| `contributor-contributor-list-module` | 7644 | 5443 | _-2201_ | _-29%_ |
| `default~code-code-example-module~code-code-tabs-module` | 57399 | 57847 | **448** | **1%** |
| `live-example-live-example-module` | 5556 | 4525 | _-1031_ | _-19%_ |
| `resource-resource-list-module` | 4838 | 3548 | _-1290_ | _-27%_ |
| `search-file-not-found-search-module` | 1603 | 1218 | _-385_ | _-24%_ |
| `toc-toc-module` | 6798 | 5579 | _-1219_ | _-18%_ |
| **_Total_** | 137439 | 124953 | _-12486_ | _-9%_ |
| &nbsp; | &nbsp; | &nbsp; | &nbsp; | &nbsp; |
| **_TOTAL_** | 765666 | 771719 | **6053** | **1%** |

</details>

---
<details open>
  <summary><b>Sizes with cli@9.0.0-next.6 <a href="https://github.com/angular/angular/pull/32923#issuecomment-537370090">with terser@4.3.2</a></b> (~3% increase with Ivy)</summary>

<p align="right"><sub><i>* Bold numbers indicate an undesirable change.</i></sub></p>

**ES2015**

| File | VE (B) | IVY (B) | Diff (B) | Diff (%) |
| --- | ---: | ---: | ---: | ---: |
| **_Eager bundles_** |
| `main` | 437453 | 456861 | **19408** | **4%** |
| `polyfills` | 52608 | 52608 | _0_ | _0%_ |
| `runtime` | 3106 | 2996 | _-110_ | _-4%_ |
| **_Total_** | 493167 | 512465 | **19298** | **4%** |
| &nbsp; | &nbsp; | &nbsp; | &nbsp; | &nbsp; |
| **_Lazy bundles_** |
| `announcement-bar-announcement-bar-module` | 2632 | 1848 | _-784_ | _-30%_ |
| `api-api-list-module` | 8357 | 6265 | _-2092_ | _-25%_ |
| `code-code-example-module` | 3120 | 2237 | _-883_ | _-28%_ |
| `code-code-tabs-module` | 35998 | 33582 | _-2416_ | _-7%_ |
| `contributor-contributor-list-module` | 7445 | 5341 | _-2104_ | _-28%_ |
| `default~code-code-example-module~code-code-tabs-module` | 52860 | 53411 | **551** | **1%** |
| `live-example-live-example-module` | 5247 | 4270 | _-977_ | _-19%_ |
| `resource-resource-list-module` | 4702 | 3460 | _-1242_ | _-26%_ |
| `search-file-not-found-search-module` | 1567 | 1213 | _-354_ | _-23%_ |
| `toc-toc-module` | 6395 | 5223 | _-1172_ | _-18%_ |
| **_Total_** | 128323 | 116850 | _-11473_ | _-9%_ |
| &nbsp; | &nbsp; | &nbsp; | &nbsp; | &nbsp; |
| **_TOTAL_** | 621490 | 629315 | **7825** | **1%** |

---
<p align="right"><sub><i>* Bold numbers indicate an undesirable change.</i></sub></p>

**ES5**

| File | VE (B) | IVY (B) | Diff (B) | Diff (%) |
| --- | ---: | ---: | ---: | ---: |
| **_Eager bundles_** |
| `main` | 482632 | 501361 | **18729** | **4%** |
| `polyfills` | 142383 | 142383 | _0_ | _0%_ |
| `runtime` | 3100 | 2990 | _-110_ | _-4%_ |
| **_Total_** | 628115 | 646734 | **18619** | **3%** |
| &nbsp; | &nbsp; | &nbsp; | &nbsp; | &nbsp; |
| **_Lazy bundles_** |
| `announcement-bar-announcement-bar-module` | 2767 | 1954 | _-813_ | _-29%_ |
| `api-api-list-module` | 9092 | 6946 | _-2146_ | _-24%_ |
| `code-code-example-module` | 3536 | 2614 | _-922_ | _-26%_ |
| `code-code-tabs-module` | 38206 | 35279 | _-2927_ | _-8%_ |
| `contributor-contributor-list-module` | 7644 | 5443 | _-2201_ | _-29%_ |
| `default~code-code-example-module~code-code-tabs-module` | 57399 | 57847 | **448** | **1%** |
| `live-example-live-example-module` | 5556 | 4525 | _-1031_ | _-19%_ |
| `resource-resource-list-module` | 4838 | 3548 | _-1290_ | _-27%_ |
| `search-file-not-found-search-module` | 1603 | 1218 | _-385_ | _-24%_ |
| `toc-toc-module` | 6798 | 5579 | _-1219_ | _-18%_ |
| **_Total_** | 137439 | 124953 | _-12486_ | _-9%_ |
| &nbsp; | &nbsp; | &nbsp; | &nbsp; | &nbsp; |
| **_TOTAL_** | 765554 | 771687 | **6133** | **1%** |

</details>

gkalpak committed:
build: avoid error in `build-packages-dist.sh`

Not sure why it works on other people's environments, but after
217db9b21 I started getting the following error when running
`scripts/build-packages-dist.sh` (on Windows):

```
ERROR: C:/.../angular/packages/bazel/docs/BUILD.bazel:3:1: Generating Skylark documentation dir for docs (3 files) failed (Exit 1)
Traceback (most recent call last):
  File "c:\...\temp\Bazel.runfiles_u_l5te\runfiles\io_bazel_skydoc\skydoc\main.py", line 335, in <module>
    main(sys.argv)
  File "c:\...\temp\Bazel.runfiles_u_l5te\runfiles\io_bazel_skydoc\skydoc\main.py", line 303, in main
    load_symbols = load_sym_extractor.extract(bzl_file)
  File "c:\...\temp\Bazel.runfiles_u_l5te\runfiles\io_bazel_skydoc\skydoc\load_extractor.py", line 110, in extract
    load_symbols = self._extract_loads(bzl_file)
  File "c:\...\temp\Bazel.runfiles_u_l5te\runfiles\io_bazel_skydoc\skydoc\load_extractor.py", line 38, in _extract_loads
    tree = ast.parse(f.read(), bzl_file)
  File "C:\...\.windows-build-tools\python27\lib\ast.py", line 37, in parse
    return compile(source, filename, mode, PyCF_ONLY_AST)
  File "packages/bazel/src/ng_package/ng_package.bzl", line 39
    print("[ng_package.bzl]", *args)
                              ^
SyntaxError: invalid syntax
```

It seems expected, because `print` is not a function, so
`print(foo, *args)` is interpreted as printing a tuple (where `*args` is
invalid syntax). Not sure why it doesn't break on other people's
machines :/

This change makes the verbose logs a little less pretty, but that
shouldn't be a big issue (given that it is an opt-in feature and it can
always be overwritten locally, if necessary).

gkalpak committed:
build(docs-infra): update current payload sizes

gkalpak committed:
build(docs-infra): turn on Ivy

The angular.io project uses Angular and CLI v9, which by default turns
on Ivy mode. However, since ec4381dd4, we explicitly opt out of Ivy.

This commit removes the `enabledIvy: false` configuration, thus allowing
the default behavior of having Ivy on.

NOTE:
This commit only changes the angular.io projects. The docs examples need
to be updated separately (first to Angular and CLI v9 and then to Ivy).

mary-poppins commented:
You can preview 7803138 at https://pr32923-7803138.ngbuilds.io/.

mary-poppins commented:
You can preview 593a7d2 at https://pr32923-593a7d2.ngbuilds.io/.

alfaproject commented on aio/scripts/_payload-limits.json:
```diff
@@ -14,22 +14,22 @@
-        "main-es5": 540416,
-        "main-es2015": 494744,
+        "main-es5": 541663,
+        "main-es2015": 496036,
```
So, it's exactly the same as Ivy? o:

gkalpak replied:
_You_ said that 😛

gkalpak replied:
IIRC, most of the lazy-loaded bundles are smaller (so the overall app size is smaller), but the main bundle is bigger with ivy (atm).

mary-poppins commented:
You can preview fc99bfb at https://pr32923-fc99bfb.ngbuilds.io/.

mary-poppins commented:
You can preview 23e0183 at https://pr32923-23e0183.ngbuilds.io/.

devversion commented on packages/bazel/src/ng_package/ng_package.bzl:
```diff
@@ -36,7 +36,7 @@ load("//packages/bazel/src/ng_package:collect-type-definitions.bzl", "collect_ty
 def _debug(vars, *args):
     if "VERBOSE_LOGS" in vars.keys():
-        print("[ng_package.bzl]", *args)
+        print("[ng_package.bzl]", args)
```
Interesting that this fails. We have the same in `rules_nodejs` [here](https://github.com/bazelbuild/rules_nodejs/blob/7f0102e2864233b789f6f916cba89a0a60375ce6/internal/linker/link_node_modules.bzl#L15-L17). Anyway, sounds reasonable to me to just pass it in like that :+1:

petebacondarwin replied:
Typo in first commit message?

> Not sure why it on other people's environments,

This phrase doesn't make sense to me

gkalpak replied:
I meant _"not sure why it **works**". Fixing...

mary-poppins commented:
You can preview d49099c at https://pr32923-d49099c.ngbuilds.io/.

gkalpak committed:
build(docs-infra): upgrade @angular/cli to 9.0.0-next.6

mary-poppins commented:
You can preview 4e1cb88 at https://pr32923-4e1cb88.ngbuilds.io/.

petebacondarwin commented on aio/tests/e2e/src/app.e2e-spec.ts:
```diff
@@ -239,9 +239,7 @@ describe('site App', function() {
-    // TODO(gkalpak): This test often times out with Ivy (because loading `guide/http` takes a lot of time).
-    //                Remove the timeout once the performance issues have been fixed.
-    }, 60000);
+    });
```
🎉

filipesilva commented:
Size difference might be related to https://github.com/terser/terser/issues/470. `terser@4.3.4` is transitively used, or at least it appears in the yarn lock file. You can try and set the yarn resolution to `4.3.2` for now. The upcoming `4.3.5` release of terser should fix the regression.

filipesilva reviewed (approved):
LGTM from dev-infra, but I think you'll want to pin `terser` using yarn resolutions to see if the payload sizes don't change. Otherwise you might have to change them back again soon.

gkalpak committed:
build(docs-infra): pin terser to 4.3.2 to avoid optimization bug

This commit pins terser to version 4.3.2 (instead of 4.3.4 which it is
currently resolved to) to avoid this bug: terser/terser#470

See https://github.com/angular/angular/pull/32923#issuecomment-537370090
for more details.

gkalpak commented:
I pinned terser to 4.3.2, as suggested by @filipesilva, and the 12KB ViewEngine regression seems to go away. I updated the numbers.

mary-poppins commented:
You can preview 907fbac at https://pr32923-907fbac.ngbuilds.io/.

gkalpak commented:
I just noticed a bug:
The main content area does not shrink to leave room for the ToC on the right (thus covering it up and essentially hiding it entirely). Investigating...

gkalpak committed:
refactor(docs-infra): replace `System.import()` with `import()`

`System.import()` has been deprecated in webpack and produces the
following warning when building the app:

```
WARNING in ./src/app/custom-elements/code/pretty-printer.service.ts
System.import() is deprecated and will be removed soon. Use import()
instead.
```

Switching to `import()` to get rid of the warning.

Fixes #30365
Closes #30419

gkalpak committed:
build(docs-infra): align config with what cli generates for new apps

This is mainly to avoid some warning when building the app, such as:

```
WARNING in .../angular/aio/src/environments/environment.archive.ts is
part of the TypeScript compilation but it's unused.
Add only entry points to the 'files' or 'include' properties in your
tsconfig.
```

(Not turning on `fullTemplateTypeCheck` due to lots of errors.)

IgorMinar reviewed (approved):
we are pretty darn close to bridging the gap for main.js! And overall (for all chunks) we are already smaller. yay!

@gkalpak once you resolve the TOC issue, this can go in! Afterwards, we should ask the team to do some manual testing on next.angular.io to confirm that there are no other regressions.

mary-poppins commented:
You can preview f8e10e0 at https://pr32923-f8e10e0.ngbuilds.io/.

gkalpak committed:
build(docs-infra): update @angular/* to 9.0.0-next.9

mary-poppins commented:
You can preview edb5bfa at https://pr32923-edb5bfa.ngbuilds.io/.

gkalpak commented:
Updating Angular to 9.0.0-next.9 fixed the ToC issue :tada:

I have added 3 new commits (since last review), with the following changes:
- ``refactor(docs-infra): replace `System.import()` with `import()` `` (to avoid warning)
- ``build(docs-infra): align config with what cli generates for new apps`` (to avoid more warnings)
- ``build(docs-infra): update @angular/* to 9.0.0-next.9`` (to fix ToC bug).

PTAL

Splaktar commented on aio/package.json:
```diff
@@ -84,17 +84,17 @@
   "dependencies": {
-    "@angular/animations": "^9.0.0-next.6",
+    "@angular/animations": "^9.0.0-next.9",
     "@angular/cdk": "8.0.0",
```
Do you plan to update Material and CDK in a separate PR? Would using `^` work?

There are some Ivy fixes in the more recent versions (and more coming soon). Latest is `8.2.2`.

gkalpak replied:
Updated. (I'll open a new PR, because I don't want to keep piggy-backing on this one :grin:)

gkalpak replied:
Created #32980.

Splaktar commented on aio/tsconfig.json:
```diff
@@ -34,6 +33,9 @@
     "tools"
```
Any plans to enable these?
```json
    "noImplicitReturns": true,
    "noFallthroughCasesInSwitch": true,
```

gkalpak replied:
Why not, I guess? The stricter the merrier :smiley:
I enabled `noFallthroughCasesInSwitch`, `noImplicitReturns` and `noUnusedParameters` :tada:

(I wish TS had a `really strict` flag that enables all strict options 😞

gkalpak replied:
Created #32980.

Splaktar commented on aio/tsconfig.json:
```diff
@@ -34,6 +33,9 @@
-  },
+    "disableTypeScriptVersionCheck": true,
+    // TODO: Fix template type errors and enable.
+    // "fullTemplateTypeCheck": true,
```
Would it be better to mark it `false` rather than comment it out?
You may also need to set `"ivyTemplateTypeCheck": false,` if doing the above.

gkalpak replied:
What would be the benefit?

Splaktar replied:
Just appeasing my ingrained dislike of commented out, version controlled code. 😁

gkalpak replied:
I am leaving a TODO comment here anyway, so there's little gain :grin:
That might even motivate you to try and fix template type-checking for aio :wink: 😆

Splaktar replied:
Maybe... I'm fixing it for material.angular.io atm 😀

gkalpak replied:
Enabled in #32980 and removed the comment.

gkalpak replied:
Ahh, I just saw your last comment. I would definitely have left it for you, had I seen it earlier 😞

Splaktar reviewed (approved):
LGTM

IgorMinar commented on aio/src/app/custom-elements/code/pretty-printer.service.ts:
```diff
@@ -26,8 +22,9 @@ export class PrettyPrinter {
     const ppo = (window as any)['prettyPrintOne'];
     return ppo ? Promise.resolve(ppo) :
-      // prettify.js is not in window global; load it with webpack loader
-      System.import('assets/js/prettify.js')
```
yay!

gkalpak replied:
:smiley:
If you liked this, you're gonna love #32980 😛

IgorMinar reviewed (approved):
lgtm! let's do this!
